### ISSUE TITLE
Allow pushing/pulling single blobs from registry

### ISSUE THREAD

**Issue by mikepurvis** [containers/skopeo#2005]
As part of using Nix to efficiently store container layers, it would be convenient to be able to download them individually, rather than only ever as part of an image. This is possible to achieve by talking directly to the registry with curl and the like, but it's fiddly to do and has terrible error-handling, particularly around the bearer token:

https://github.com/nlewo/nix2container/blob/80babc1470efe3561d5000b6f822b95b4f15df35/default.nix#L148-L157

Would Skopeo consider adding functionality to the CLI to operate at the layer level as well as on whole images? At its simplest, this could be a flag on the existing copy verb that selects the desired blob by hash (or maybe a list of them, if there's a use-case for that?), so like:

```
$ skopeo copy docker://library/alpine dir:. --specific-blob 8a49fdb3b6a5ff2bd8ec6a86c05b2922a0f7454579ecc07637e94dfd1d0639b6
Copying blob 8a49fdb3b6a5 done

$ ls
8a49fdb3b6a5ff2bd8ec6a86c05b2922a0f7454579ecc07637e94dfd1d0639b6
```

I don't know enough about registries to know if the reverse operation of copying a single, unassociated blob to one is an operation that makes much sense.

Anyway, thanks for considering it. If this isn't a good fit for the mainline Skopeo CLI, I'd love a pointer on how to best implement a minimal wrapper of my own that can call into the internals of Skopeo or other libraries to achieve this functionality in a way that will be supportable going forward.

**Comment by mtrmac** [comment]
Thanks for reaching out.

That single-layer access really feels rather out of scope for Skopeo itself; it _is_ available in the underlying containers/image library as a stable API (e.g. `docker.NewReference()` → `.NewImageSource()` → `.GetBlob()`), so building a specialized Go program that does whatever you want that way would be the general recommendation.

---

Actually, reading a bit more of the surrounding Nix code, AFAICS the code doesn’t just want to fetch a _single_ blob, but all of them, _and_ the config, i.e. basically the whole image. Why not use `skopeo copy docker://… dir:` to do all of that, then? That would even fetch the layers in parallel; and possibly allow enforcing signature policy; _and_ validate the integrity of the data vs. the digests, to detect corruption or malicious registries.

You obviously know about `dir:`, so I wonder what I’m missing. Is the manifest itself somehow modified from the on-registry one?

---

(There also is an undocumented Skopeo subcommand that… could be found. But it has no known users, and it’s not something we expect to document or ever improve. It could serve as an example of the c/image API use, though.)

**Comment by mikepurvis** [comment]
Thanks for looking and for the thoughts!

> ... reading a bit more of the surrounding Nix code, AFAICS the code doesn’t just want to fetch a single blob, but all of them, and the config, i.e. basically the whole image. Why not use `skopeo copy docker://… dir:` to do all of that, then?

Understandable— there is a bit of a limitation here on the Nix side that I can't do a single derivation (a package/buildable) that outputs multiple fixed-output store paths, so to get the layer blobs each in their own store path requires pulling them each in their own derivation. (`buildImageBlob` is a function returning the result of a `pkgs.runCommand`, which is a derivation that builds the store path for that blob)

> That would even fetch the layers in parallel

We actually still get the parallism because Nix builds those individual blob derivations in parallel— but it's one better: since those are fixed-output Nix store paths that are _also_ content-addressed, if I download five different images, it'll only grab that base layer once, whereas a `skopeo copy docker://… dir:` call would of course fetch the base layers repeatedly since it has no knowledge of what already exists in other directories.

> it is available in the underlying containers/image library as a stable API (e.g. `docker.NewReference()` → `.NewImageSource()` → `.GetBlob()`)

Fantastic, I'll investigate this route. Thanks again and feel free to close this ticket.
